### PR TITLE
feat: New fallback logic for geometry in analytics [2.40-DHIS2-13528-2]

### PR DESCRIPTION
**Pull request by d-bernat**
Fallback logic:
there are three variables: coordinateField, fallbackCoordinateField and defaultCoordinateFallback. 

coordinateField is clear,  fallbackCoordinateField is hard fallback. When fallbackCoordinateField is null and defaultCoordinateFallback is true (default), a cascade fallback is applied (psigeometry -> pigeometry -> teigeometry -> ougeometry). Teigeometry is valid only for registration programs (or registered?).

This behavior is valid only for event tables (enrollment ones have only pigeometry).

### CODE REVIEW

**Review by maikelarabori — COMMENTED** [review]
Nice work @d-bernat. I just left a few comments.

Cheers,
Maikel

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/EventQueryParams.java [hunk 1]
```diff
@@ -244,15 +231,11 @@
...
+     * order defines priority of geometry fields.
      */
-    private String coordinateField;
-
-    /**
-     * The fallback coordinate field to use as basis for spatial event
-     * analytics, SQL COALESCE applied on coordinate fields.
-     */
-    private String fallbackCoordinateField;
+    private List<String> coordinateFields;
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/EventQueryParams.java:238]: discussed, closed

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/EventDataQueryService.java [hunk 2]
```diff
@@ -71,23 +73,15 @@
...
-
-    /**
-     * Returns the coordinate column field to use for the given coordinate
-     * field. Coordinate field must match EVENT, a data element identifier or an
-     * attribute identifier.
-     *
-     * @param coordinate the coordinate field.
-     * @param program the program uid.
+     * @param program the coordinate field.
+     * @param coordinateField the coordinate field.
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/EventDataQueryService.java:77]: Done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/EventDataQueryService.java [hunk 3]
```diff
@@ -71,23 +73,15 @@
...
+     * @param fallbackCoordinateField the coordinate field.
+     * @param defaultCoordinateFallback the coordinate field.
      * @return the coordinate column field.
-     * @throws IllegalQueryException if the coordinate field is not valid.
+     * @throws IllegalQueryException if any coordinate field is not valid.
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/EventDataQueryService.java:81]: accepted

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java [hunk 4]
```diff
@@ -408,80 +421,120 @@ public EventQueryParams getFromAnalyticalObject( EventAnalyticalObject object )
...
     }
 
+    /**
+     * Returns list of coordinateFields.
+     *
+     * All possible coordinate fields are collected. The order defines the
+     * priority of geometries and is used as a paramaters in sql coalesce
+     * function
+     *
+     * @param program the coordinate field.
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java:431]: done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java [hunk 5]
```diff
@@ -408,80 +421,120 @@ public EventQueryParams getFromAnalyticalObject( EventAnalyticalObject object )
...
+            .addAll( getFallbackCoordinateFields( program, fallbackCoordinateField, defaultCoordinateFallback ) );
+
+        return coordinateFields.stream().distinct().collect( Collectors.toList() );
+    }
+
+    // -------------------------------------------------------------------------
+    // Supportive methods
+    // -------------------------------------------------------------------------
+
+    private boolean verifyFallbackCoordinateField( boolean isRegistration, String coordinateField )
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java:500]: Done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java [hunk 6]
```diff
@@ -408,80 +421,120 @@ public EventQueryParams getFromAnalyticalObject( EventAnalyticalObject object )
...
+     *
+     * All possible coordinate fields are collected. The order defines the
+     * priority of geometries and is used as a paramaters in sql coalesce
+     * function
+     *
+     * @param program the coordinate field.
+     * @param coordinateField the coordinate field.
+     * @param fallbackCoordinateField the coordinate field.
+     * @param defaultCoordinateFallback the coordinate field.
+     * @return List<String>
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/DefaultEventDataQueryService.java:435]: done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import org.hisp.dhis.common.ValueType;
+import org.hisp.dhis.feedback.ErrorCode;
+
+/**
+ * @author Dusan Bernat
+ */
+public interface EventCoordinateService
+{
+    /**
+     * verified the validity of fallback coordinate field
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java:42]:
```suggestion
     * verifies the validity of fallback coordinate field
```

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManagerTest.java [hunk 8]
```diff
@@ -741,4 +788,32 @@
         queryItem.setFilters( new ArrayList<>( filters ) );
         return queryItem;
     }
+
+    private EventQueryParams getEventQueryParamsForCoordinateFieldsTest( List<String> coordinateFields,
+        boolean explicit )
```
[the remark this answers is not in the document]
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManagerTest.java]: done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @param valueType value type
+     * @param field geometry
+     * @param errorCode code for standard error message
+     * @return verified geometry or
+     */
+    String getCoordinateFieldOrFail( ValueType valueType, String field, ErrorCode errorCode );
+
+    /**
+     *
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java:74]: I think you missed the method description and the return description.

**Review by maikelarabori — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java (hunk 7 above)
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java]: done

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java (hunk 9 above)
Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/EventCoordinateService.java:75]: done

**Review by gnespolino — APPROVED** [review]

**Pull request merged**